File: packages/cli-node/src/process-utils.ts
// Found via real end-to-end testing (Tasks.md 2.2 verification notes), not
// spec-driven: on Windows, spawning with `shell: true` runs the command
// through cmd.exe, so child.pid is cmd.exe's PID — child.kill() only kills
// the shell wrapper, leaving the actual server process (e.g. a `npx`-spawned
// node process) running and its stdio pipes open. `taskkill /T` kills the
// whole process tree instead.
import { spawnSync } from "node:child_process";
import type { ChildProcess } from "node:child_process";

export function killProcessTree(child: ChildProcess): void {
  if (child.pid === undefined) return;
  if (process.platform === "win32") {
    spawnSync("taskkill", ["/pid", String(child.pid), "/T", "/F"], { stdio: "ignore" });
  } else {
    child.kill();
  }
}

File: packages/cli-python/mcpseal/process_utils.py
# Mirrors packages/cli-node/src/process-utils.ts's finding (Tasks.md 2.2
# verification notes): on Windows, spawning through a shell means the
# tracked PID is the shell wrapper's, not the real server process's, so a
# plain terminate()/kill() leaves the actual child running with its stdio
# pipes open. `taskkill /T` kills the whole process tree instead.
import subprocess
import sys

# Spawn MCP servers through a shell ONLY on Windows.
#
# On Windows the shell is required to resolve the .cmd/.bat shims that
# npm installs (`npx`, `npx.cmd`), and Popen joins the argument list into
# a command line via list2cmdline, so [command, *args] survives intact.
#
# On POSIX, `Popen([command, *args], shell=True)` does NOT do that: it
# runs `/bin/sh -c "<command>"` and demotes the remaining elements to
# $0/$1/..., so the server is spawned with NO arguments. A stdio server
# started that way never speaks JSON-RPC (e.g. bare `python` sits reading
# source from stdin), so the proxy blocks forever on its first response
# read. POSIX needs shell=False, where the list is exec'd directly.
USE_SHELL = sys.platform == "win32"


def kill_process_tree(proc: subprocess.Popen) -> None:
    if proc.pid is None:
        return
    if sys.platform == "win32":
        subprocess.run(
            ["taskkill", "/pid", str(proc.pid), "/T", "/F"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    else:
        try:
            proc.kill()
        except ProcessLookupError:
            pass
